Lifecycle Events V1 [1.15]

i509VCB opened the pull request:
1.16 Sister PR is #681

Now includes Chunk and (Block)Entity (un)load events as well.

Creates a new module `fabric-lifecycle-events-v1`

This also creates `fabric-item-api-v1` for the `ItemTooltipCallback` that doesn't belong in the lifecycle module.

Legacy events have been deprecated and verified that they still work.

(Un)Load events have been tested, but still converting them to fabric's testing format and cleaning up some of the tracking mechanism in testing.

### **Edit 1:**
Lifecycle events have been moved to either `ServerXYZEvents` or `ClientXYZEvents` depending on whether the event fires on a logical client or server. Note events defined in `ClientXYZEvents` are client only.

**Renamed Events:**
- `ServerStartCallback` -> `ServerLifecycleEvents.SERVER_STARTED`
- `ServerStopCallback` -> `ServerLifecycleEvents.SERVER_STOPPING`
- `ServerTickCallback` -> `ServerTickEvents.END_SERVER_TICK`
- `WorldTickCallback`: This has been split up to correspond with the logical world type.
  - `ServerWorld` -> `ServerTickEvents.END_WORLD_TICK`
  - `ClientWorld` -> `ClientTickEvents.END_WORLD_TICK`

- `ClientTickCallback` -> `ClientTickEvents.END_CLIENT_TICK`
- `ItemTooltipCallback` -> Moved to `fabric-item-api-v1`

**New Events (Client & Server):**
- Entity Load
  - `ClientEntityEvents.LOAD_ENTITY`
  - `ServerEntityEvents.LOAD_ENTITY`

- BlockEntity Load
  - `ClientBlockEntityEvents.LOAD_BLOCK_ENTITY`
  - `ServerBlockEntityEvents.LOAD_BLOCK_ENTITY`

- BlockEntity Unload
  - `ClientBlockEntityEvents.UNLOAD_BLOCK_ENTITY`
  - `ServerBlockEntityEvents.UNLOAD_BLOCK_ENTITY`

- Chunk Load
  - `ClientChunkEvents.LOAD_CHUNK`
  - `ServerChunkEvents.LOAD_CHUNK`

- Chunk Unload
  - `ClientChunkEvents.UNLOAD_CHUNK`
  - `ServerChunkEvents.UNLOAD_CHUNK`

- Ticking
  - `ClientTickEvents.START_CLIENT_TICK`
  - `ClientTickEvents.START_WORLD_TICK`

**New Events (Server)**
- Server Stopped
  - `ServerLifecycleEvents.SERVER_STOPPED` (Called the instant after the server has completely stopped. This is the last event called on a dedicated server before the process closes.)

- Ticking
  - `ServerTickEvents.START_SERVER_TICK`
  - `ServerTickEvents.START_WORLD_TICK`

**New Events (Client)**
- `ClientEntityEvents.UNLOAD_ENTITY` (This is client only since tracking entity unloads on a server is too unreliable).
- `ClientLifecycleEvents.CLIENT_STARTED` (Called when the client is about to tick for the first time)
- `ClientLifecycleEvents.CLIENT_STOPPING` (Called when the client begins to stop. (Called before the client's player is disconnected if in game))

__A few more thing to do:__
 ~~- Add tick events to correspond to the beginning of a tick on a client, server and world~~
 ~~- Docs. Lots of them~~

i509VCB committed:
Lifecycle Events V1

Now includes Chunk and (Block)Entity (un)load events

i509VCB committed:
Add some tests to verify worlds are ticking in the log

LemmaEOF commented on fabric-generic-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/generic/v1/item/ItemTooltipCallback.java:
```diff
@@ -0,0 +1,48 @@
+	/**
+	 * Fired after the game has appended all base tooltip lines to the list.
+	 */
+	Event<ItemTooltipCallback> EVENT = EventFactory.createArrayBacked(ItemTooltipCallback.class, callbacks -> (stack, context, lines) -> {
```
Should we still have this in its own separate `EVENT` class, since it looks like we're going in the direction of having classes that define lots of different events?

i509VCB replied:
Then again, a majority of the events in the API still follow that `EVENT` rule. Due to the fact there are references to client types, I couldn't easily do a `EntityUnloadCallback.CLIENT/SERVER` without classloader issues. The other route was double the amount of interfaces in the api

liach replied:
Imo for that you can do `EntityUnloadCallback` and `ClientEvents.ENTITY_UNLOAD` where
`EntityUnloadCallback ENTITY_UNLOAD = EventFactory.createArrayBacked(xxx);` in `ClientEvents` class

liach replied:
Also how is this a "generic event"?

i509VCB replied:
Yes, it is not really a generic event. But the item tooltip stuff is surely not lifecycle by any metric.

I'd say it makes more sense in an `item` module but this is temporary

liach commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java:
```diff
@@ -0,0 +1,196 @@
+	});
+
+	public static final Event<BlockEntityUnloadCallback<ClientWorld>> BLOCK_ENTITY_UNLOAD = EventFactory.createArrayBacked(BlockEntityUnloadCallback.class, callbacks -> (blockEntity, world) -> {
+		final Profiler profiler = world.getProfiler();
```
any point of getting the profiler before the profiler enabled check?

i509VCB replied:
Really nothing. Just wanted to throw the profiler into the variable stack and forget about it. I'll probably move it into the if block

i509VCB replied:
All the profiler calls should be within the `isProfilingEnabled` block now

liach commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -0,0 +1,204 @@
+
+	public static final Event<WorldTickCallback<ServerWorld>> WORLD_TICK = EventFactory.createArrayBacked(WorldTickCallback.class, callbacks -> world -> {
+		if (EventFactory.isProfilingEnabled()) {
+			final Profiler profiler = world.getProfiler();
```
yes, you do it like this, only get profiler after the isProfilingEnabled check

liach commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleCallback.java:
```diff
@@ -0,0 +1,26 @@
+ * A callback which is invoked when the server advances in it's lifecycle.
+ */
+public interface ServerLifecycleCallback {
+	void onChangeLifecycle(MinecraftServer server);
```
Maybe `lifecycle`?

i509VCB replied:
Well this means the server's lifecycle stage has changed. onLifecycle is a bit vague it would somewhat correspond to a `lifecycle` method

liach replied:
`lifecycle` is a verb. It has `cycle` meaning it is cycling to a next stage.

Pyrofab replied:
Sorry what ? https://www.merriam-webster.com/dictionary/life%20cycle

liach replied:
Gradle has a `lifecycle` that is like a `info` for logging. This current name is acceptable.

JamiesWhiteShirt commented on fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/impl/event/lifecycle/client/LegacyClientEventInvokers.java:
```diff
@@ -0,0 +1,34 @@
+		ClientLifecycleEvents.CLIENT_TICK.register(ClientTickCallback.EVENT.invoker()::tick);
+		ClientLifecycleEvents.WORLD_TICK.register(WorldTickCallback.EVENT.invoker()::tick);
+		// This is part of generic events now.
+		net.fabricmc.fabric.api.client.event.generic.v1.item.ItemTooltipCallback.EVENT.register(ItemTooltipCallback.EVENT.invoker()::getTooltip);
```
The invokers of all of the above legacy events are being captured here, which is invalid. That is, for each event here, the value of `EVENT.invoker()` is only evaluated in this method, where the callback is being registered for the replacement event, and not each time the new event is invoked. The Event class allows the value of `invoker()` to change when new listeners are registered.

For the array-based events which all of fabric API uses, the invoker is updated each time a new listener is registered, which means this will fail to notify any listeners that are registered after this initializer is called.

This will have to be changed to evaluate the value of `invoker()` each time the event is called, probably with an explicit lambda instead of a method reference.

i509VCB replied:
Changed it to a lambda.

shedaniel commented on fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/impl/event/lifecycle/client/LegacyClientEventInvokers.java:
```diff
@@ -0,0 +1,34 @@
+	public void onInitializeClient() {
+		// Allows deprecated events to still be invoked by the newer implementations
+		ClientLifecycleEvents.CLIENT_TICK.register(ClientTickCallback.EVENT.invoker()::tick);
+		ClientLifecycleEvents.WORLD_TICK.register(WorldTickCallback.EVENT.invoker()::tick);
```
do we already tick twice per tick for worlds? once for server and once for client?

i509VCB replied:
Yes.

JamiesWhiteShirt commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/impl/event/lifecycle/LifecycleInternalListeners.java:
```diff
@@ -0,0 +1,46 @@
+					ServerLifecycleEvents.BLOCK_ENTITY_UNLOAD.invoker().onUnloadBlockEntity(blockEntity, world);
+				}
+			}
+		});
```
We might want to impose a stricter order for when the SERVER_STOP and ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD events are called. Let's say a mod registers a listener for both SERVER_STOP and ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD. Depending on whether the mod gets to register its SERVER_STOP event listener _before_ or _after_ this listener (which may be nondeterministic), it will have its SERVER_STOP event listener called either before or after all the ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD events are invoked.

I imagine a mod might use these events to free resources, and might want to use some server-bound resource which is freed when the SERVER_STOP event is called when entities and block entities are unloaded. Having all the entities and block entities unload before the server is unloaded would make the most sense.

i509VCB replied:
I could fire that internal listener a bit earlier so all entities would be ublosded before the server stops. However some event listeners may already use the fact not every entity is unloaded during server shutdown themselves, so we would be shooting them in the foot then. I could hook this into the server world being closed so it is garunteed the world no longer will exist, and some what covers custom worlds added by other mods

i509VCB replied:
@JamiesWhiteShirt So I decided to make this case be bound to the closing of the server worlds so events can still work off the old contract where entities were still present on the server

shedaniel reviewed (commented):
I don't really understand the concept of 'generic' events, it sounds super vague to me.

Also might want to add Javadocs explaining where the event field is and what it does in their callback class.es

i509VCB committed:
Lambda boogaloo

i509VCB committed:
Add some docs. Distinguish between a server starting to stop and server which has stopped.

i509VCB committed:
Split up test mods, some tweaks to (block)entity (un)load events.

Bind the ServerWorld being closed during shutdown to unload (block)entities.

i509VCB committed:
Shift around a few profiler variables and finalize

Prospector commented:
Fixes #47

i509VCB commented:
Well part of #47, well dont have world unload calkbacks.

Prospector commented:
Hm true. Probably best not making this PR any bigger unless they're really easy :P

i509VCB commented:
Funnilly enough the chunk events are the most reliable in my testing

liach commented:
Would you mind incorporating https://github.com/FabricMC/fabric/pull/445 into this pr?

i509VCB commented:
I'll look into it.

i509VCB committed:
Complete the tests, Block entities on server should be reliably tracked now.

Entities on the server obviously still need to be wrangled.

i509VCB committed:
Drop Server Entity Unload callback.

Believe me, this was a hard decision but it stands on the fact that about 20-40% of entities silently unload without going through the proper "unloadEntity" method in ServerWorld. No amount of debug hacks, double tracking unload events and even replacing the entity maps do not fix this issue. So I have decided to drop this from the feature set.

i509VCB committed:
checkstyle

i509VCB commented:
So I had to drop the server entity unload callback. This is because 20-40% of entities silently unload without going through the proper "unloadEntity" method in ServerWorld. No amount of debug hacks, double tracking unload events and even replacing the entity maps do not fix this issue. The only fix is a major rewrite of how mojang tracks entities in the game.

Other than that this is complete and tested.

i509VCB commented:
@liach would like you to ponder this question:
> Do we want to use a `fabric-item-events` or `fabric-item-api` module instead of `generic-events`

liach commented:
In fact, this is more like a client event, as tooltip is shown only on client in screens. So may be screen/client/render event category

i509VCB commented:
Yes, that is true it is client only. The issue with a `fabric-client-events` module is that is very broad. I think one of the `item` named modules would be better since it specifically refers to what the event actually targets.

I also understand this is a question of, would there be events of a similar type in the future, thereby either making the module name make more or less sense.

i509VCB committed:
generic-events -> item-api

i509VCB commented:
Generic events is now `item-api` since the only event there currently is related to items. And I do feel there will be some more item related stuff in the future that could belong there.

YanisBft commented:
Shouldn't the item tooltip callback be in the rendering api module actually?

i509VCB commented:
Nope, only lets you add to the tooltip, gives you no rendering capabilities.

i509VCB committed:
Server start -> Server started

i509VCB committed:
Allow getting current server from Lifecycle

People have asked for this, but it is not encouraged for obvious reasons. Should be staged well enough to revert if we decide to.

i509VCB commented:
Considering whether adding a WorldLoad and unload event (for server) would make sense in this PR here or be a scope creep. Also due to 1.16, things may not be 100% stable there.

i509VCB committed:
checkstyle lol

i509VCB committed:
update injection name

i509VCB committed:
Checkstyle lol: Redundant modifiers

i509VCB commented:
Last travis build failed due to network stuff on their end. Compiles on my end

i509VCB committed:
Add client starting, stopping and stopped callbacks.

i509VCB committed:
Loicenses

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Called when a client ticks.
```
These all need more specific documentation, the various ticks all last quite a while. It may even be necessary to have separate events for start and end, e.g. end of world tick may be used to start async computations for the next tick.

i509VCB replied:
Added some slight documentation about where the tick events are called. (There are START and END callbacks now)

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java:
```diff
@@ -0,0 +1,235 @@
+	});
+
+	/**
+	 * Called when a chunk is loaded into a ClientWorld.
```
For load/unload the timing is crucial and should be explained. Ideally all load events occur after the subject has been added to the world/chunk with the world+chunk being normally usable and obtainable via MinecraftServer. This ensures that e.g. an area search can be done without fragile postponing workarounds. Unload should happen while still reachable for similar reasons.

Additionally, access to the de-/serialization phases would be useful such that custom data can be embedded in the vanilla nbt tree as it gets read or written. It is useful for persistent data attachments to chunk or world, e.g. to store worldgen metadata for retrogen implementations.

i509VCB replied:
I've had a discussion on discord related to a `Save/LoadEntityDataCallback` and the concenus was it isn't very useful unless we have a way to attach data to an entity (via something like CCA) or mixin to said entity to access the data you wish to store.

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java:
```diff
@@ -0,0 +1,235 @@
+		}
+	});
+
+	public static final Event<GameLifecycleCallback<MinecraftClient>> CLIENT_STARTING = EventFactory.createArrayBacked(GameLifecycleCallback.class, callbacks -> client -> {
```
These lack documentation, not sure what they are supposed to do?

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/GameLifecycleCallback.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * A callback which is invoked when the game's engine advances in it's lifecycle.
+ */
+public interface GameLifecycleCallback<T> {
```
I'd use two instances without generics for ease of use. It may be worthwhile to organize the events+callbacks differently, maybe an outer class like ClientTickEvents that holds both the event fields and callbacks as nested classes? Similar for others, ClientLifecycleEvents is getting big and the many trivial top level callback interfaces are getting a bit too spread out.

i509VCB replied:
I've split up the `Client`/`ServerLifecycleEvents` into more specific classes.

Because of the split, at most these more specific classes (So there is a `ClientTickEvents.World` and `ServerTickEvents.World`) now only hold 2-3 inner classes for the interfaces for event listeners implement.

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/GameTickCallback.java:
```diff
@@ -0,0 +1,21 @@
+package net.fabricmc.fabric.api.event.lifecycle.v1;
+
+public interface GameTickCallback<T> {
+	void onTick(T game);
```
Same with this generic parameter, also avoids implementing e.g. chunk+world tick handlers in the same class.

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -0,0 +1,246 @@
+	});
+
+	/**
+	 * Called when the server has started. At this stage, all worlds are live.
```
".. and it is about to tick for the first time" maybe? assuming that's correct?

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -0,0 +1,246 @@
+	});
+
+	/**
+	 * Called when the server has started stopping. All worlds are still present.
```
This could clarify what stopping means, especially with the client having a less permanent stop (in a way).

Player3324 commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -0,0 +1,246 @@
+	 * @return the currently running server
+	 * @throws IllegalStateException if the server is not available
+	 */
+	public static MinecraftServer getCurrentServer() {
```
While this is useful and sometimes not truly avoidable, I am not sure if we want to reinforce the notion that there's only one server. There should at least be a strong recommendation to always use world.getServer and other context specific approaches instead.

i509VCB replied:
Yea that is precisely why I did stage this addition to one commit incase it was unpopular. The documentation would be nice to have about reccomending getting the server from the ServerWorld.

Internally this will optionally get the depending on a client or server. Which you can look at in impl packages.

i509VCB replied:
Hmm, I think this might `getPrimaryServer` might be a better name since we don't assume one server exists. I've added some docs explaining how it's use is discouraged.

i509VCB committed:
Reorganize so each event category has it's own class.

Also this collapses the pretty widely reaching interfaces into more specific inner classes to avoid issues with generics.

i509VCB committed:
Some docs and slight name changes

i509VCB committed:
Add start tick callbacks to worlds, server and client

liach commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -61,17 +61,17 @@ private ServerLifecycleEvents() {
+	 * Gets the currently running primary server.
 	 *
-	 * <p><b>Use of this method is highly impractical and not recommended since there is no real restriction on whether the game engine could run multiple servers concurrently.</b>
+	 * <p><b>Use of this method is highly discouraged and not recommended since there is no real restriction on whether the game engine could run multiple servers concurrently.</b>
```
just do `@Deprecated`

i509VCB committed:
Enhance some client related docs to life cycle

i509VCB committed:
Deprecate for reasons of discouraging singletonish server getter methods in lifecycle

i509VCB committed:
Add some description related to integrated server on server stopping.

i509VCB committed:
Add small test to verify tick starts are right spot of load

i509VCB committed:
Docs and a tiny bit more testing.

i509VCB committed:
Try clarifying client docs

i509VCB committed:
Drop a slightly unnessecary event

i509VCB committed:
Actually call and implement START_SERVER_TICK event

i509VCB committed:
Remove non-existent test

i509VCB committed:
again

i509VCB committed:
Refer to minecraft itself in client lifecycle docs

i509VCB committed:
Refer to Minecraft itself within ServerLifecycleEvents

i509VCB commented:
I have opened a sister PR for 1.16 #681

i509VCB commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java:
```diff
@@ -0,0 +1,107 @@
+	 * @throws IllegalStateException if the server is not available
+	 */
+	@Deprecated
+	public static MinecraftServer getPrimaryServer() {
```
After thinking a bit, I am going to drop these methods.

liach replied:
Please do.

i509VCB replied:
And it's out

i509VCB committed:
Remove primary server getters

i509VCB committed:
IJ DO YOU SPEAK RESOLVING IMPORTS

(cherry picked from commit c9257e8a11d8361469349f4171263121bb111af7)

i509VCB committed:
Prune the tests that shouldn't exist

i509VCB committed:
Listen here checkstyle you bugger

(cherry picked from commit 9701bba4002cec089c9d3738b1f226128078c130)

i509VCB committed:
Split up events to individual interfaces. Make Chunk events use WorldChunk instead.

Player3324 reviewed (approved):
LGTM, thanks!

liach commented on fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java:
```diff
@@ -33,9 +33,9 @@ private ClientLifecycleEvents() {
-			callback.onChangeLifecycle(client);
+	public static final Event<ClientStarted> CLIENT_STARTED = EventFactory.createArrayBacked(ClientStarted.class, callbacks -> client -> {
+		for (ClientStarted callback : callbacks) {
+			callback.onClientStarted(client);
```
`clientStopping.onClientStopping` etc sounds weird 🤔

i509VCB replied:
yea it definitely felt odd to type out. Any better ideas for name?

modmuss50 reviewed (commented):
Taken a quick scan read of it, and it seems very good.

I will do a full review shorly.

modmuss50 commented on fabric-item-api-v1/build.gradle:
```diff
@@ -0,0 +1,6 @@
+archivesBaseName = "fabric-item-api-v1"
```
Not sure if we really need a whole module just for `ItemTooltipCallback`

Prospector replied:
To me, it makes sense. It gives a home for future item hooks. Better than a huge `misc` module that would become massively bloated as time goes on.

i509VCB replied:
Like prospector said, this is a thought of, what would likely exist in the future where this one can live

modmuss50 reviewed (approved):
Looks good, just a bit unsure on the 1 api, module.

Geometrically commented:
What are the advantages of these compared to normal callbacks?

Ueaj-Kerman commented:
they look pretty normal to me

i509VCB commented:
The callbacks still exist, the actual `Event<XYZCallback>` fields all exist in specific classes so packages aren't nearly as wide.
